include/swift/AST/Expr.h

```diff
@@ -3650,6 +3650,43 @@ class PackExpansionExpr final : public Expr {
   }
 };
 
+/// An expression to materialize a pack from a tuple containing a pack
+/// expansion, spelled \c tuple.element.
+///
+/// These nodes are created by CSApply and should only appear in a
+/// type-checked AST in the context of a \c PackExpansionExpr .
+class MaterializePackExpr final : public Expr {
+  /// The expression from which to materialize a pack.
+  Expr *FromExpr;
+
+  MaterializePackExpr(Expr *fromExpr, bool implicit, Type type)
+      : Expr(ExprKind::MaterializePack, implicit, type),
+        FromExpr(fromExpr) {}
+
+public:
+  static MaterializePackExpr *create(ASTContext &ctx, Expr *fromExpr,
+                                     bool implicit = false,
```

Comment:
If you make “implicit” last parameter CSApply could skip it.

Reply:
I am also curious why this is not implicit by default if it could only be synthesized by the compiler?

Reply:
It's not implicit because it's written explicitly in source code as `tuple.element`. The `.element` part corresponds to `MaterializePackExpr`, but the constraint system only knows that after an overload choice has been bound for the member. I should probably add an `ElementLoc` here since the `FromExpr` will always just be the base tuple value.

Reply:
Adding ElementLoc is a good idea. I just thought that we set implicit to everything that is compiler generated/transformed it allows to skip location validation as well but in this case it doesn’t matter because the locs are used from the original node.